Migration to Navigatum

PiotrKedra opened the pull request:
## Issue

This fixes the following issue(s): 
- Resolves: #1451
- Resolves: #1456 
- Resolves: #1338
- Resolves: #415
- Resolves: #1182

Depending on merging centralized search #1457, I decided to work on this PR rather than the master branch because I used some room finder logic there.

## Description
Migrate to:
- more information nav.tum.sexy/api/get/:id api-endpint 
- the search nav.tum.sexy/api/search?q=querry api-endpoint

PiotrKedra commented:
Migration more or less should be ready. I have migrated three roomfinder endpoints: 
- fetchAvailableMaps()
- fetchRooms()
- fetchCoordinates()

Into two NavigaTum endpoints (search navigation and navigation details)

### What is still to be done?
1. remove old roomfinder logic (didn't remove it yet because of the below problems)
2. besides migration I want to slightly improve the navigation details UI (now it only displays sitemap plan), to sth like on the NavigaTum website. Also, I want to allow to open location in other apps #1456.

### Problems/questions
1. I heard that NavigaTum is not stable yet, therefore can it be merged now?
2. Old roomfinder API has an endpoint that fetches RoomFinderSchedule. However, I couldn't make this code run in the App, so I am not sure if this is still supported or it can be removed?
3. Barrier-free API fetches rooms that have ID from the old roomfinder API and using NavigaTum with this ID does not work. Therefore, now when users click on the Barrier-free facility, the app redirects to the search screen with this facility address as a query input. Is this solution ok?

CommanderStorm commented:
> 1. I heard that NavigaTum is not stable yet, therefore can it be merged now?
If we comment on https://github.com/TUM-Dev/navigatum/issues/130 about the change timeline, it is stable ;)

> 2. Old roomfinder API has an endpoint that fetches RoomFinderSchedule. However, I couldn't make this code run in the App, so I am not sure if this is still supported or it can be removed?
@joschahenningsen historically What is the status on this? Was this ever used in the app?

> 3. Barrier-free API fetches rooms that have ID from the old roomfinder API and using NavigaTum with this ID does not work. Therefore, now when users click on the Barrier-free facility, the app redirects to the search screen with this facility address as a query input. Is this solution ok?

@joschahenningsen Can we migrate the backend (add a new field?)

joschahenningsen commented:
> @joschahenningsen historically What is the status on this? Was this ever used in the app?

I think this might be used in the calendar somewhere. I'll need to do some more digging though.

> @joschahenningsen Can we migrate the backend (add a new field?)

Sure, as long as it does not break old clients :) Do you have a way of mapping those ids? 

On another note: Do y'all need some help with the merge conflicts? :D

PiotrKedra commented:
## RoomFinderSchedule

I did some digging into this secret RoomFinderSchedule logic ;p
> 2. Old roomfinder API has an endpoint that fetches RoomFinderSchedule. However, I couldn't make this code run in the App, so I am not sure if this is still supported or it can be removed?

I find out that this logic was not available to execute after these two commits:
- [moved to Picasso on Apr 27, 2018](https://github.com/TUM-Dev/Campus-Android/commit/f03650e9654ea677123392ca28b444e5dca78a60#diff-c3a8aa040517eb62a994a98d9e6cb86d4aeb851a1bb64162c948d65f6be3d065)
- [Update various dependencies on Oct 13, 2020](https://github.com/TUM-Dev/Campus-Android/commit/5236de171266a6b79b5c24ce68fdbcfe12bf19a5#diff-1c4f58e643463b7977c699461a8ae3db68743711905b3fa0ad1f3037d4b3fe4c)

In both these commits, it looks like this might be removed accidentally, but cannot be sure. However, it looks like, since April 2018 this RoomFinderSchedule feature does not work, so I think it would be best to deal with this problem in a separate issue.  Also since it was not available for so long, there is a question if anybody needs it.

## Barrier-free API

Another case is what shall we do with this Barrier-free API
> 3. Barrier-free API fetches rooms that have ID from the old roomfinder API and using NavigaTum with this ID does not work. Therefore, now when users click on the Barrier-free facility, the app redirects to the search screen with this facility address as a query input. Is this solution ok?

Should we wait for https://github.com/TUM-Dev/Campus-Backend/pull/80 to be merged and ready to use, so it can be implemented here as well? Or we just merge this PR now with my temporary solutions and take care of Barrier-free in another one? @CommanderStorm

CommanderStorm commented:
> Should we wait for [TUM-Dev/Campus-Backend#80](https://github.com/TUM-Dev/Campus-Backend/pull/80) to be merged and ready to use, so it can be implemented here as well? Or we just merge this PR now with my temporary solutions and take care of Barrier-free in another one? @CommanderStorm

Actually, we don't need to wait for that PR, as the current backend does already provide us with the exact information we need:
![image](https://user-images.githubusercontent.com/26258709/180622708-d95d4d4e-cef2-4d10-95e2-ffc812985667.png)
When looking in the network tab, `room_code` is the ID, NavigaTUM uses, so probably we can simplify this UX a bit.  
The study rooms also one place where we already have this field. (it is named `code` in the data class)
Thus, we can redirect directly to showing the room.

Doing so should cover all cases, since [the fixes we added](https://github.com/TUM-Dev/navigatum/blob/main/data/sources/15_patches-rooms_tumonline.yaml) for invalid `room_code`s do not relate to study rooms or barierfree. Redirect to the details page for the search results sould be save here..

However, if you like, we can use your solution and migrate all occurrences of the `room_id` to `room_code` in a later PR…

PiotrKedra commented:
Thanks for finding the room_code parameter @CommanderStorm. I have implemented it and it works as expected. I have also removed not used room finder code and merged master. There were lots of conflicts, but after merging all seems to work (I hope 😅).

From my side, I think PR is ready to be merged.  I will create a separate issue for the RoomFinderSchedule feature as described earlier.

CommanderStorm commented on app/src/main/java/de/tum/in/tumcampusapp/component/ui/search/adapter/RecentSearchesAdapter.kt:
```diff
@@ -60,25 +60,36 @@ class RecentSearchesAdapter(
+                        val lecture = Lecture.fromRecent(recentSearch)
+                        titleText.text = lecture.title
                     } catch (exception: Exception) {
                         titleText.setText(R.string.not_available_search)
                     }
-                    icon.setImageResource(R.drawable.ic_room)
+                    icon.setImageResource(R.drawable.ic_lecture)
                     icon.setBackgroundResource(R.drawable.search_result_icon_background)
                     profilePicture.setImageDrawable(null)
                 }
-                RecentsDao.LECTURES -> {
+                RecentsDao.NAVIGA_TUM_BUILDINGS -> {
                     try {
-                        val lecture = Lecture.fromRecent(recentSearch)
-                        titleText.text = lecture.title
+                        val navigationEntity = NavigationEntity.fromRecent(recentSearch)
+                        titleText.text = navigationEntity.getFormattedName()
                     } catch (exception: Exception) {
                         titleText.setText(R.string.not_available_search)
                     }
-                    icon.setImageResource(R.drawable.ic_lecture)
+                    icon.setImageResource(R.drawable.ic_building)
+                    icon.setBackgroundResource(R.drawable.search_result_icon_background)
+                    profilePicture.setImageDrawable(null)
+                }
+                RecentsDao.NAVIGA_TUM_ROOMS -> {
+                    try {
+                        val navigationEntity = NavigationEntity.fromRecent(recentSearch)
+                        titleText.text = navigationEntity.getFormattedName()
+                    } catch (exception: Exception) {
+                        titleText.setText(R.string.not_available_search)
+                    }
+                    icon.setImageResource(R.drawable.ic_room)
                     icon.setBackgroundResource(R.drawable.search_result_icon_background)
                     profilePicture.setImageDrawable(null)
```
Since this is 2x basically the same code except for the setImageResource calls, we should simplify this and have an interiour when clause for this

CommanderStorm commented on app/src/main/java/de/tum/in/tumcampusapp/component/ui/search/SearchFragment.kt:
```diff
@@ -135,8 +140,9 @@ class SearchFragment : BaseFragment<Unit>(
                 RecentsDao.PERSONS -> openPersonDetails(Person.fromRecent(recent))
-                RecentsDao.ROOMS -> openRoomDetails(RoomFinderRoom.fromRecent(recent))
                 RecentsDao.LECTURES -> openLectureDetails(Lecture.fromRecent(recent))
+                RecentsDao.NAVIGA_TUM_BUILDINGS -> openNavigationDetails(NavigationEntity.fromRecent(recent))
+                RecentsDao.NAVIGA_TUM_ROOMS -> openNavigationDetails(NavigationEntity.fromRecent(recent))
```
```suggestion
                RecentsDao.NAVIGA_TUM_BUILDINGS, RecentsDao.NAVIGA_TUM_ROOMS -> openNavigationDetails(NavigationEntity.fromRecent(recent))
```

CommanderStorm replied:
```suggestion
                RecentsDao.NAVIGA_TUM_BUILDINGS, RecentsDao.NAVIGA_TUM_ROOMS -> openNavigationDetails(NavigationEntity.fromRecent(recent))
```

CommanderStorm commented on app/src/main/res/values/strings.xml:
```diff
@@ -771,18 +771,35 @@ Signature: %5$s</string>
+    <string name="interactive_map">Interactive map</string>
+    <string name="open_location_in_other_app">Open location in other app</string>
+    <string name="available_maps">Available site plans:</string>
+    <string name="map">Map</string>
```
map does not seem to be used anywhere else. Is this intentional?

PiotrKedra replied:
My mistake, I should have remove it

PiotrKedra replied:
Ah, it is used in fragment_navigation_details spinner ^^

CommanderStorm commented on app/src/main/res/values-de/strings.xml:
```diff
@@ -719,14 +719,32 @@ Signatur: %5$s</string>
     <string name="clear_search_history_request">Möchten Sie den Suchverlauf löschen?</string>
 
+    // Navigation details
+    <string name="location_details">Location details</string>
+    <string name="loading_details">loading</string>
+    <string name="site_plans">Site plans</string>
+    <string name="interactive_map">Interactive map</string>
+    <string name="open_location_in_other_app">Open location in other app</string>
+    <string name="available_maps">Available maps:</string>
+    <string name="map">Map</string>
```
```suggestion
    <string name="location_details">Details zum Ort</string>
    <string name="loading_details">lädt</string>
    <string name="site_plans">Lagepläne</string>
    <string name="interactive_map">Interaktive Karte</string>
    <string name="open_location_in_other_app">Ort in anderer Anwendung öffnen</string>
    <string name="available_maps">Verfügbare Karten:</string>
    <string name="map">Karte</string>
```

CommanderStorm commented on app/src/main/java/de/tum/in/tumcampusapp/api/app/TUMCabeClient.java:
```diff
@@ -280,24 +279,12 @@ public List<BuildingToGps> getBuilding2Gps() throws IOException {
                 .body();
     }
 
+    @Deprecated // use NavigaTUM api instead
```
```suggestion
    @Deprecated // This API has been deprecated. Use the equivalent NavigaTUM API instead
```

CommanderStorm reviewed (approved):
Thanks for all of this work. 😄 
I have commented on some small things inline. Once these are resolved, IMO we can merge this. 

Given the size of this PR, I would like somebody else to double-check me @Bentipa, would you have the time to look through this PR?

CommanderStorm reviewed (approved):
Thanks for all of this work. 😄 
I have commented on some small things inline. Once these are resolved, IMO we can merge this. 

Given the size of this PR, I would like somebody else to double-check me @Bentipa, would you have the time to look through this PR?

Bentipa commented:
> Thanks for all of this work. 😄 
> 
> I have commented on some small things inline. Once these are resolved, IMO we can merge this. 
> 
> 
> 
> Given the size of this PR, I would like somebody else to double-check me @Bentipa, would you have the time to look through this PR?

Hey,

as you may have noticed I did not have time yet to also answer/review the other things as I am sick at the moment. Although I will take a look once I am back, if you need a review for this one faster we should maybe ask someone else.

PiotrKedra commented on app/src/main/java/de/tum/in/tumcampusapp/api/navigatum/domain/NavigationEntity.kt:
```diff
@@ -0,0 +1,54 @@
+        return removeHighlight(subtext)
+    }
+
+    private fun removeHighlight(field: String): String {
```
Actually, I haven't thought of it much. Mb one concern is that in centralized search there are ppl and lectures fetched as well and they do not have any highlights, but I am not sure if this is wrong. I think we can add highlights it just needs additional logic that handles it.

PiotrKedra commented on app/src/main/java/de/tum/in/tumcampusapp/api/navigatum/model/details/RoomFinderMapDto.kt:
```diff
@@ -0,0 +1,15 @@
+data class RoomFinderMapDto(
+    var id: String = "",
+    var name: String = "",
+    @SerializedName("file")
```
I think it can stay like it is. Didn't know that there are multiple sizes ;d

PiotrKedra commented on app/src/main/java/de/tum/in/tumcampusapp/component/other/locations/LocationManager.kt:
```diff
@@ -311,16 +262,20 @@ class LocationManager @Inject constructor(c: Context) {
         if (loc.contains("(")) {
             loc = loc.substring(0, loc.indexOf('('))
-                    .trim { it <= ' ' }
+                .trim { it <= ' ' }
```
Yeah probably u are right, but I think I didn't touch these lines, they were the same before ;p

PiotrKedra commented on app/src/main/res/layout/toolbar_search.xml:
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="utf-8"?>
-<com.google.android.material.appbar.AppBarLayout xmlns:android="http://schemas.android.com/apk/res/android"
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
We don't handle these gestures here I guess, but if we maybe want in future, then I think your way is better. And yeah, I did it due to runtime error, didn't come up with your solution. I guess I need to improve my skills ;D

PiotrKedra commented:
Thanks for the extensive review :) WIth most of the comments I agree. I think it would be better If I aggregate these changes into one commit, rather than one by one by clicking here on GitHub (there will be long history ;d)? But I don't know what is better, I see these commit suggestions for the first time ^^

When it comes to ordering since there are four different objects fetched in a centralized search. The "All results" sort alphabetically all of them, but when you choose a specific object type, then the ordering is as it was fetched from APIs. I added sorting because I thought it was slightly better to mix the results alphabetically than display groups of objects one by one (firstly all people, then all lectures, then all buildings and rooms). This was my reasoning, what do you think? :D

CommanderStorm commented:
> Thanks for the extensive review :) WIth most of the comments I agree. I think it would be better If I aggregate these changes into one commit, rather than one by one by clicking here on GitHub (there will be long history ;d)? But I don't know what is better, I see these commit suggestions for the first time ^^

What works for you…
You don't have to worry about a long history, since all changes are squashed into one commit anyway 😉 

> When it comes to ordering since there are four different objects fetched in a centralized search. The "All results" sort alphabetically all of them, but when you choose a specific object type, then the ordering is as it was fetched from APIs. I added sorting because I thought it was slightly better to mix the results alphabetically than display groups of objects one by one (firstly all people, then all lectures, then all buildings and rooms). This was my reasoning, what do you think? :D

For rooms/buildings, this definitively does not make sense, as we have a priority between our items.
For people and lectures I dont know this for certain, but I assume they are also sorted somehow.

We should therefore just drop this sorting. Finding a total order between these items is not trivial and should be its PR

From an ordering, I would therefore prefer:
- people
- building/room (in the order we communicate if possible)
- lectures

Assumptions being,
- that there are few human names that map intersect with buildings/rooms and therefore this can be first without causing much of a bad impact.
- Searching for your lectures is something you rarely do in a campus app

PiotrKedra commented:
Ok, I will fix the ordering as u suggest, but probably I will do it next week after my exams

Bentipa commented:
I will take another look once @PiotrKedra has implemented the changes, just tag me again then please :)

octycs commented on app/src/main/java/de/tum/in/tumcampusapp/api/navigatum/domain/NavigationEntity.kt:
```diff
@@ -0,0 +1,54 @@
+    var type: String = "",
+    @SerializedName("name")
+    var name: String = "",
+    @SerializedName("subtext")
+    var subtext: String = ""
+) : Serializable {
+
+    fun getFormattedName(): String {
+        return removeHighlight(name)
+    }
```
```suggestion
    @SerializedName("subtext")
    var subtext: String = "",
    @SerializedName("parsed_id")
    var parsedId: String? = null,
) : Serializable {

    fun getFormattedName(): String {
        return if (parsedId === null) {
            removeHighlight(name)
        } else {
            removeHighlight(parsedId!!) + " ➤ " + removeHighlight(name)
        }
    }
```

I think it would be helpful to add the `parsed_id` field as well. This way it would look like this:

<img src="https://user-images.githubusercontent.com/7429408/183260283-399bfb7e-5ada-408a-93a9-0c93392eb6fc.png" width="400" />

We could also use a different arrow symbol, e.g. `»`:

<img src="https://user-images.githubusercontent.com/7429408/183260328-e966abb5-063d-4355-93fc-190dda137f2c.png" width="400" />

If you think this makes the search results too long, I would suggest showing only the `parsed_id`, because it's only included if we consider it more helpful than the normal name.

(You can also change the sources, I've never written Kotlin so far)

PiotrKedra replied:
I have added ➤ option

PiotrKedra commented on app/src/main/java/de/tum/in/tumcampusapp/component/tumui/roomfinder/NavigationDetailsFragment.kt:
```diff
@@ -0,0 +1,211 @@
+            val adapter = ArrayAdapter<String>(requireContext(), android.R.layout.simple_spinner_item)
+            adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
+            binding.availableMapSpinner.adapter = adapter
+            availableMaps.forEach {
```
Spinner shows when the details object is being loaded. There is no spinner for loading the map itself.

PiotrKedra commented on app/src/main/java/de/tum/in/tumcampusapp/component/tumui/roomfinder/NavigationDetailsFragment.kt:
```diff
@@ -0,0 +1,211 @@
+    }
+
+    private fun loadMapImage(map: NavigationMap) {
+        val pinDrawable = ResourcesCompat.getDrawable(resources, R.drawable.ic_place, null)
```
I have changed it to tum_orange

PiotrKedra commented:
I have implemented changes, and also left comments on some conversations
@Bentipa

CommanderStorm reviewed (approved):
imo this can be merged. All my major comments have been adressed.
@Bentipa do you want to review it, or should I merge it?

CommanderStorm commented:
@PiotrKedra Thanks for the wonderful work you did ✨

Bentipa commented:
@CommanderStorm Just gave it a quick look cause of time reasons, but looks good overall to me. So we can merge it.
